src/mono/mono/metadata/icall.c

```diff
@@ -6307,9 +6317,9 @@ ves_icall_RuntimeType_make_array_type (MonoReflectionTypeHandle ref_type, int ra
 {
 	MonoType *type = MONO_HANDLE_GETVAL (ref_type, type);
 
-	MonoClass *klass = mono_class_from_mono_type_internal (type);
-	check_for_invalid_array_type (klass, error);
+	check_for_invalid_array_type (type, error);
 	return_val_if_nok (error, MONO_HANDLE_CAST (MonoReflectionType, NULL_HANDLE));
+	MonoClass *klass = mono_class_from_mono_type_internal (type);
```

Comment:
I wonder if it is better to pass `MonoClass *klass` to `check_for_invalid_array_type`, since line 6298 will always be executed. That way we could save one function call.

Reply:
So the problem is that if you start with a `MonoType` and then call `mono_class_from_mono_type_internal` it is not an exact roundtrip.  In particular you use the `MonoType:byref` bit.  Because a MonoClass is not able to represent a byref type.  (This is kind of an annoying limitation in Mono but fixing it is probably a ton of work).

I thought about instead returning the `klass` from `check_for_invalid_array_type`, but it's kind of a weird incidental return value.  In any case `mono_class_from_mono_type_internal` is pretty quick to call since it's just a switch on a value (`MonoType:type`) that we already looked at.

Reply:
Gotcha! My original thought was to returning the `klass` from `check_for_invalid_array_type`. But since it is not costly, I am okay with current implementation.